tools/manifestfile.py: Add a library for working with manifests.

jimmo opened the pull request:
Work in progress... comments/thoughts welcome!

This is the first step towards updating MicroPython's package management and deployment system in a way that combines freezing, on-device-install and host-based-install.

This PR does seven things:
 - Splits a library to process manifest files out of makemanifest.py 
 - Extends manifests to be more about "I require this package/module", and make freezing a specific case of that.
 - Simplify the common case of defining a Python module or package (this is especially the case for libraries in micropython-lib and drivers from the main repo).
 - Simplify common uses of the include() function ("/manifest.py" is implied if the path is a directory).
 - Make micropython-lib dependencies able to be expressed with `require()` rather than include. This will later generalise to use cases where the resource is fetched from the network rather than having the repo checked out locally.
 - ~Update all manifests in this repo to use the new features.~  (Moved to #9037)
 - Add a Python wrapper for mpy-cross (based on https://gitlab.com/alelec/mpy_cross by @andrewleech), and update makemanifest.py to use it. This will be used to generate pre-compiled binaries for modules to be published for on-device installation.

Part of this is inspired by #[8231](https://github.com/micropython/micropython/pull/8231) (@andrewleech) and the idea is that this "manifestfile" library can be used by tools like mpremote to manage a project and its dependencies that can be deployed to a board. Similarly, the mpy-cross library can be used by mpremote to compile "on-the-fly" during deployment.

In more detail, the idea is to use this to support the following use case:
```
$ mpremote connect /dev/ttyACM0 install foo
```
where `foo` is a package from micropython-lib, and "foo" and its dependencies will be installed on the device. For more sophisticated uses-cases:
```
$ mpremote connect /dev/ttyACM0 deploy app_manifest.py
```
where the whole app and dependencies (as described by module/package/include/require commands in the manifest) will be deployed in an rsync-like process. Additionally, it would integrate with #8381 in order to build memory-mappable images -- i.e. the benefits of freezing, but without recompiling or even needing to access the main repo at all).

The next step after this PR is to update the process that builds https://micropython.org/pi (this is where upip downloads resources from) to use this new library, and the subsequently update upip.py to work with this. The result will be:
 - upip starts working again
 - upip will download compiled .mpy files rather than .py files

There are some details to still sort out, most importantly the future of PyPI in this picture. Supporting PyPI from upip is quite complicated. Producing libraries that can be deployed to PyPI and deploy correctly via upip is actually quite difficult and unnecessarily burdensome for third-party developers, and quite inefficient from the device side. Randomly sampling a few projects confirmed that this is causing problems.

Personally I would prefer to see more libraries added to micropython-lib, but for most projects that are a single file or a package of a few files I would rather invest the effort in having upip be able to install directly from the author's GitHub repo  e.g. `>>> upip.install('github:foo/bar')` where the repo contained some simple descriptor/metadata file that was generated from a manifest.py. In many cases, a single-file library with no dependencies could be `>>> upip.install('github:foo/bar/bar.py')`. The same commands would work from `mpremote` as described above too.

jimmo committed:
tools/manifestfile.py: Add library for working with manifests.

This splits the manifest file loading logic from makemanifest.py and
updates makemanifest.py to use it.

This will allow non-freezing uses of manifests, such as defining packages
and dependencies in micropython-lib.

Also adds additional methods to the manifest "API":
 - require() - to get a package from micropython-lib.
 - module() - to define a single-file module
 - package() - to define a multi-file package

module() and package() should replace most uses of freeze() and can also
be also used in non-freezing scenarios.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
mpy-cross/mpy_cross: Add Python wrapper for mpy-cross.

Rather than invoking mpy-cross directly via system/subprocess in our build
tools and other tools, this provides a Python interface for it.

Based on https://gitlab.com/alelec/mpy_cross (with the intention of
eventually replacing that as the "official" pypi distribution once setup.py
etc are added).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/makemanifest.py: Update to use mpy_cross module.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/manifestfile.py: Allow include of directory path.

If an include path is a directory, then it implicitly grabs the manifest.py
file inside that directory. This simplifies most manifest.py files.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Note: The CI will fail as this requires the corresponding PR to micropython-lib (https://github.com/micropython/micropython-lib/pull/506 to add manifest.py for all libraries) to be merged first.

dlech commented:
> * upip will download compiled .mpy files rather than .py files

It seems like we would need something like https://peps.python.org/pep-0427/#file-name-convention to avoid problems when breaking MPY ABI version changes are made since there will be a transition period when people will be running both old and new firmwares.

jimmo commented:
> > * upip will download compiled .mpy files rather than .py files
> 
> It seems like we would need something like https://peps.python.org/pep-0427/#file-name-convention to avoid problems when breaking MPY ABI version changes are made since there will be a transition period when people will be running both old and new firmwares.

Yes. In addition, the native arch matters too for any library that uses `@native` or `@viper` (or potentially if we add some dynamic native modules to micropython-lib). Right now the .mpy file includes a two-byte identifier that includes the mpy version and architecture. Encoding those two bytes into the resource request / name should be enough to ensure a compatible .mpy.

The more interesting part is how to ensure that all possible valid .mpy versions are generated for each library. My proposal would be to always produce the current latest version with the current mpy-version (for all arches), and never delete old builds. In other words, a board running mpy-version=8 could run any micropython-lib library version that was either current when the mpy-version moved to 8, or any subsequent version released until mpy-version moves to 9. This would be easy to extend to n previous versions if necessary (i.e. we always run current _and_ previous mpy-cross).

andrewleech commented:
Thank you @jimmo ! A heap of with has been done here and it sounds fantastic, it sounds very aligned with some idealistic fantasies I had about the manifest system :-D

Just on the note about mpy versions, I've had a number of requests with mpy-cross package to easily support distributing of versions, both in issue tracker and on slack.

While the two bytes are the formal identifier I think it would really help users to have more human readable identifiers of the bytecode-version_arch in the name.

There's likely to also be more requests come in to easily build old bytecode versions - not sure of the cleanest way of doing this. That can be a future todo.

I'll lol over this pair of PR's in more detail in the coming days, cheers!

andrewleech commented:
Another thought, it might be worth having another "field" in the mpy package name format for something akin to ROM_LEVEL so you could have a min, extra, full version of packages? There's a number of stdlib packages on micropython-lib that I go back and forth on adding particular functions, trying to trade off small vs useful, it would be great to just have certain python files in core package, them plus the rest in an extra package.

aivarannamaa commented:
> There are some details to still sort out, most importantly the future of PyPI in this picture. Supporting PyPI from upip is quite complicated. Producing libraries that can be deployed to PyPI and deploy correctly via upip is actually quite difficult and unnecessarily burdensome for third-party developers, and quite inefficient from the device side. Randomly sampling a few projects confirmed that this is causing problems.

I assume you mean the process of creating the special tar.gz-s is too complicated.

Have you thought of allowing 3rd party libs to publish regular source-dists to PyPI and providing a proxy index at micropython.org, which does the conversion and outputs tar.gz files suitable for upip?

I have implemented the opposite process in [pipkin](https://pypi.org/project/pipkin/), which uses a temporary localhost proxy for presenting upip-compatible sdists at PyPI as regular sdists so that they can be installed with pip.

IMO, ditching PyPI and standard Python packaging approach because of upip is not a good idea. I have written more about this here: https://github.com/aivarannamaa/pipkin#current-state-and-goals

andrewleech commented:
> Have you thought of allowing 3rd party libs to publish regular source-dists to PyPI and providing a proxy index at micropython.org, which does the conversion and outputs tar.gz files suitable for upip?

The idea of an on-the-fly repacking proxy to pypi is interesting, however I'm still struggling to see the real value of this. A new tool as described here that can go directly to GitHub/gitlab/url for third party publishers, cutting out the middle-man of pypi/proxy seems so much simpler to me. At this point I can't see much extra value the pypi provides? 

If pypi is directly exposed on-device via upip we will forever get issue reports from people trying to install micropython-foo on their device without realising it's for circuitpython instead, or a desktop tool for cpython.

Pypi lacks the abi/arch/platform/bytecode version specifiers available for micropython tools to ever properly check for compatibility from metadata like pip can, nor can we push micropython binary packages.

Sure, maybe it would be possible to work closer with pypa to get things like this added, but this will likely be a lot of work for someone to organise.

> I have implemented the opposite process in [pipkin](https://pypi.org/project/pipkin/), which uses a temporary localhost proxy for presenting upip-compatible sdists at PyPI and micropython.org/pi as regular sdists so that they can be installed with pip.
> 
> IMO, ditching PyPI and standard Python packaging approach because of upip is not a good idea. I have written more about this here: https://github.com/aivarannamaa/pipkin#current-state-and-goals

Thanks for the link, I have only briefly trialled pipkin myself and hadn't seen this page. It all feels like a _lot_ of work to bend pip and friends to our needs. I've been working with extending pip/setuptools in different ways for many years on cpython projects with things like nuitka, briefcase, my mpy-cross package, pip-system-certs, wrapt, etc. The biggest thing they have in common is a lot of fragile magic that often breaks. And as pip evolves (eg pyproject.toml) pretty much all these tools are broken and need a lot of work to rewrite almost from scratch to integrate into the new pip/setuptools.

I truly believe that attempting to bend our tools to use pypi as a glorified free fileserver will end up costing more and more maintenance over time.

aivarannamaa commented:
> Pypi lacks the abi/arch/platform/bytecode version specifiers available for micropython tools to ever properly check for compatibility from metadata like pip can, nor can we push micropython binary packages.

There is [PEP 425](https://peps.python.org/pep-0425/), which allows different compatibility tags for wheels, including custom Python tag (_"Other Python implementations should use `sys.implementation.name`"_). I have experimented with `micropython` tag myself. I could successfully upload [wheels using it to PyPI](https://pypi.org/project/mp-packaging-experiment/#files) and later [install them with pip using `--implementation` argument](https://github.com/aivarannamaa/packaging_experiment). So, unless I misunderstood you, the framework for maintaining non-CPython packages on PyPI is already available. Unfortunately tools are not there yet (for example `pip install --implementation` does not affect dependency selection), but I believe this can be fixed, either by new tools, hacks like pipkin or by asking for changes to pip.

> I truly believe that attempting to bend our tools to use pypi as a glorified free fileserver will end up costing more and more maintenance over time.

IMO the free fileserver part is not the most important thing here. It's the common standards and common package namespace, which allow flexible development of packages and tools. It is not hard to imagine a library which is useful dependency in a MicroPython, CircuitPython and CPython library or app, but such combination would be more difficult to achieve if MicroPython packages lived in their own separate world.

Even with a new approach optimized for MicroPython, one would still need to think about some kind of standards, at least when 3rd party contributions are desired. If you also want to allow some overlap with CPython or CircuitPython world, the task of creating a new system is not so simple anymore.

romilly commented:
First reaction: I really want this (or something close to it) to happen. I've a dozen small projects I'd love to package, with more on the way. Some would benefit from native code components.

jimmo commented:
This has been rebased. I have split the actual changes to the manifest.py files into a separate PR (#9037) because of the ordering of the changes needs:
- This PR
- Update micropython-lib to add manifest.py for all libraries (https://github.com/micropython/micropython-lib/pull/506)
- Update the micropython-lib submodule
- Update main repo manifest.py files (#9037)

jimmo committed:
top: Use micropython-lib unconditionally in manifests.

micropython-lib is now a submodule, and the manifest compilation process
will ensure it is available, so manifests no longer need to check that it
is available.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/manifestfile.py: Allow require() to specify unix packages.

By default, don't include micropython-lib/unix-ffi in the search.

If unix_ffi=True is passed to require(), then include unix-ffi and make it
take precedence over the other locations (e.g. python-stdlib).

This does two things:
 - Prevents non-unix builds from using unix-only packages.
 - Allows the unix build to optionally use a more full-featured (e.g. ffi)
   based package, even with the same name as one from e.g. stdlib.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Thanks @andrewleech and @aivarannamaa for the discussion.

> Another thought, it might be worth having another "field" in the mpy package name format for something akin to ROM_LEVEL so you could have a min, extra, full version of packages? There's a number of stdlib packages on micropython-lib that I go back and forth on adding particular functions, trying to trade off small vs useful, it would be great to just have certain python files in core package, them plus the rest in an extra package.

Yes, for sure. The `aioble` package kind of does this right now via the options that you can pass to `include()` (or now `require()`). See the [Nano Connect board defn] (https://github.com/micropython/micropython/blob/master/ports/rp2/boards/ARDUINO_NANO_RP2040_CONNECT/manifest.py#L4) which uses this. But that only works for freezing right now.

I have a couple of other thoughts on this:
 - Rather than the aioble way of selectively including which files, it would be interesting to add a way to load "constants" into mpy-cross, such that you could essentially "preprocess" out sections of code in a single file (via the optimiser dropping `if False:`).
 - If we end up implementing options 2 or 4 from #9018, then we will have a mechansim for providing composable modules. i.e. you could install "foo" and optionally "foo_feature1" and "foo_feature2", but still end up with just "import foo".

> I assume you mean the process of creating the special tar.gz-s is too complicated.

Pretty much. My main point here is that the argument for PyPi / pip is "it's the Python way" but as soon as you're in the territory of doing custom stuff, then a lot of that appeal disappears. The main thing you have to get right is making sure that the destination paths are correct. The other concern I had is that currently we need to set a custom window size, but also we can't guarantee that all of our ports will be able to do https to pypi.org if they ever upgrade their certs.

> Have you thought of allowing 3rd party libs to publish regular source-dists to PyPI and providing a proxy index at micropython.org, which does the conversion and outputs tar.gz files suitable for upip?

Initially I considered running proxy service that essentially MITMed the request to PyPI. But we really don't want to commit to running any infrastructure. The option you're describing sounds more like doing this statically -- fetch all "micropython" packages and build a static service for this. But I guess I come back to...what's the point of PyPi in that scenario? We end up scooping up all our packages to deploy them to PyPi only to download them again in another form. Why not just build the target directly?

> So, unless I misunderstood you, the framework for maintaining non-CPython packages on PyPI is already available. Unfortunately tools are not there yet (for example wheel doesn't allow specifying custom implementation tag and pip install --implementation does not affect the dependency selection), but I believe this can be fixed, either by new tools, hacks like pipkin or by asking for changes in pip.

I definitely agree that it would be possible, but it just seems like so much extra work to do something that is fundamentally so simple. I really don't want to be chasing the leading edge of new features in the tools and pip.

> It's the common standards and common package namespace

The PyPi namespace is a bit of a mess. We don't have control of many of the micropython-xyz packages (even the ones that are from micropython-lib). It's also not a useful place to go searching for packages, it's very hard to get useful results.

> It is not hard to imagine a library which is useful dependency in a MicroPython, CircuitPython and CPython library or app

I'm less convinced on a "micro" package being useful for CPython although can't rule it out I guess, but definitely can imagine the MicroPython+CircuitPython combo.

I guess there are three scenarios for such a package:
 - The package lives in micropython-lib
 - The package lives in one of the CircuitPython bundles  (so it's also in pypi if it's not in the community bundle)
 - The package lives... somewhere else.   (maybe in pypi, up to the maintainer)

But the thing here is that PyPi/pip isn't part of CircuitPython's workflows either, except for the Blinka/Linux use case and now Thonny. To my knowledge, pip/pypi are not part of Adafruit's plans for their web & ble workflows (i.e. via code.circuitpython.org), which as far as I can tell will access their bundles directly.

I'm not against doing what Adafruit do and publishing all micropython-lib libraries to PyPi just as a matter of course (with no custom anything, just the .py files directly). We can generate pyproject.toml from manifest.py and do this automatically (modulo issues with claiming all the package names on PyPi). This will open up the micropython-lib packages to Thonny/pipkin.

I guess that's the summary: happy for more sophisticated tools/workflows to be built on top of this, but I'm trying to work towards the simplest possible solution that addresses use cases that cover the vast majority of users:
- describe source packages from (main repo, micropython-lib, user repo) that will be frozen
- on-device (wifi/ethernet) or mpremote (or tool using mpremote as a library): install a package by name from micropython-lib
- on-device or mpremote: install a package by URL from arbritrary hosting (e.g. github)
- mpremote: deploy a manifest to a device
- use arch-specific .mpy files in all cases

Our manifest.py mechanism will be part of this (because it's a necessary part of freezing), so this PR is a step towards making that system easier to use and more flexible. We considered an non-programatic alternative to the manifest.py format (e.g. manifest.json or manifest.toml) but the Python approach is a good mix of flexibility and ability to run the build without any additional Python dependencies. (And it's kind of an implementation detail anyway -- anything outside this repo will use the output of the manifest compilation which is run automatically, whether that be the static web backend for upip or pyproject.toml etc).

jimmo committed:
tools/manifestfile.py: Allow manifests to set metadata.

The metadata can be version, description, and license.

After executing a manifest, the top-level metadata can be queried, and also
each file output from the manifest will have the metadata of the
containing manifest.

Use the version metadata to "tag" files before freezing such that they have
__version__ available.

aivarannamaa commented:
Thank you @jimmo for this explanation!

> I'm not against doing what Adafruit do and publishing all micropython-lib libraries to PyPi just as a matter of course [...]

I think this is a good idea even if it doesn't seem necesssary right now. I'm aware of the difficulties with some micropython-lib PyPI names, but IMO it's worth securing at least the other names, even if just for avoiding more confusion in the future. It seems to be common (and accepted) practice of registering names at PyPI even if the packages get published elsewhere. BTW, Adafruit recently agreed to keep publishing packages to PyPI to make pipkin's life easier: https://github.com/thonny/thonny/issues/2352#issuecomment-1192111600.

> I guess that's the summary: happy for more sophisticated tools/workflows to be built on top of this, [...]

I'm happy to hear this!

Maybe I missed it, but are there plans to add any runtime support for querying the metadata of installed packages? Those more sophisticated tools would need to know the current state of packages for given device. For built-ins (and frozens?) there is help("modules"), for others one can list the /lib directory but it would be great if it was possible to know the PyPI (or micropython-lib) names and versions of the installed packages. That's why pipkin keeps a minimal .dist-info directory next to each installed package, but currently it has no means for getting this info for frozen packages. I imagine it could be possible to present a read-only directory containing .dist-info directories corresponding to frozen packages, but a built-in function returning name-version pairs would do as well.

I have a good example for this issue from CircuitPython world -- CP builds for most boards have frozen adafruit-circuitpython-busdevice, but for some you need to install it separately. Currently pipkin needs to do heuristics on help("modules") output in order to decide, whether to install it or not as a dependency of another package. It would be much cleaner if it could query the dist names and versions of the frozen packages.

You mentioned the promotion of mpy files. The package metadata would be also relevant for development tools like linters and autocompleters, which could then automatically download corresponding source code or typing stubs. More simply -- it would be good to know what's available on a MicroPython device.


Another topic I haven't seen beeing discussed (or I simply haven't understood) -- according to the new plan, how will one specify dependencies of a package?

andrewleech commented:
> according to the new plan, how will one specify dependencies of a package?

Take a look at some of the proposed matching changes in micropython-lib like base64: https://github.com/micropython/micropython-lib/pull/506/files#diff-9d23e8ab880a19b4c369d64a6b68f23c40db71e0aeef4e6980092d7ef4d992b7

> It would be much cleaner if it could query the dist names and versions of the frozen packages.

I think at this stage there's no plan to have metadata copied into frozen code / installed on device.
I too think this is worth adding in future, even if it's just optional. In many user cases maintaining a minimum footprint is most important, however in many other use cases having the traceability the metadata could/should give would be an enormous benefit.
Even if they're not installed on-device, perhaps it'd be easy enough to cache / store them in the build folder alongside the elf/bin files, this would be great for checking licences etc.

On that note @jimmo it'd probably be worth adding a licence kwarg to the metadata command (even if it's still just a no-op for now)

romilly commented:
In the medium term I'd find version and license metadata very valuable.

On Fri, 12 Aug 2022 at 10:25, Andrew Leech ***@***.***> wrote:

> according to the new plan, how will one specify dependencies of a package?
>
> Take a look at some of the proposed matching changes in micropython-lib
> like base64:
> https://github.com/micropython/micropython-lib/pull/506/files#diff-9d23e8ab880a19b4c369d64a6b68f23c40db71e0aeef4e6980092d7ef4d992b7
>
> It would be much cleaner if it could query the dist names and versions of
> the frozen packages.
>
> I think at this stage there's no plan to have metadata copied into frozen
> code / installed on device.
> I too think this is worth adding in future, even if it's just optional. In
> many user cases maintaining a minimum footprint is most important, however
> in many other use cases having the traceability the metadata could/should
> give would be an enormous benefit.
> Even if they're not installed on-device, perhaps it'd be easy enough to
> cache / store them in the build folder alongside the elf/bin files, this
> would be great for checking licences etc.
>
> On that note @jimmo <https://github.com/jimmo> it'd probably be worth
> adding a licence kwarg to the metadata command (even if it's still just a
> no-op for now)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython/pull/8914#issuecomment-1212910663>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABEOPWMHQHPO4W454W2QBLVYYKAFANCNFSM53VYZZPQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
***@***.*** ***@***.*** skype:romilly.cocking web:
http://blog.rareschool.com/

aivarannamaa commented:
> Take a look at some of the proposed matching changes in micropython-lib like base64: https://github.com/micropython/micropython-lib/pull/506/files#diff-9d23e8ab880a19b4c369d64a6b68f23c40db71e0aeef4e6980092d7ef4d992b7

Do I understand properly that the connecting parts are `require("binascii")` in python-stdlib/base64/manifest.py and `module("binascii.py")` in python-stdlib/binascii/manifest.py? Or the `require("binascii")` simply refers to folder python-stdlib/binascii?

How would it work for dependencies outside of python-stdlib?

jimmo commented:
> Maybe I missed it, but are there plans to add any runtime support for querying the metadata of installed packages?

I haven't written about this anywhere, but yes this has come up a few times. The simple answer is that I would like `mod.__version__` to work, and I plan to add support to manifestfile.py to inject this automatically during the freezing and upip-deploy process. My understanding is that the CircuitPython bundler does something like this too.

I updated the micropython-lib PR yesterday with the WIP support for this. There's now a `metadata` command that allows the manifest to specify the version for the package that this manifest describes. See for example https://github.com/micropython/micropython-lib/pull/506/files#diff-355585a8ac48343d1df93a9a0adb50e0c7ee3f67509e5bda15d44d7c42391842
(Sorry haven't had a chance to push the corresponding change to this PR yet)

> Do I understand properly that the connecting parts are require("binascii") in python-stdlib/base64/manifest.py and module("binascii.py") in python-stdlib/binascii/manifest.py? Or the require("binascii") simply refers to folder python-stdlib/binascii?

The latter. micropython-lib packages are uniquely identified by their leaf directory name.

`module("binascii.py")` command literally means "this manifest uses binascii.py as a module". (As opposed to the `package` command which pulls in a directory).

> How would it work for dependencies outside of python-stdlib?

This works across all of micropython-lib. i.e. python-ecosys can require packages python-stdlib.

But the more interesting scenario is for example a manifest in a standalone repo (i.e. a board definition) that is either used for freezing or with the `mpremote deploy app_manifest.py` scenario described above. In that case if the depenency is another git repo available locally, then `include()` is what you want. For arbitrary dependencies (i.e. in a non-local github repo) this is the purpose of the (not yet implemented) `FILE_TYPE_HTTP` in this PR. i.e. I plan to support `require("github:user/repo")`.

aivarannamaa commented:
> The next step after this PR is to update the process that builds https://micropython.org/pi (this is where upip downloads resources from) to use this new library, and the subsequently update upip.py to work with this.

Would you continue with current naming scheme, ie. `micropython-` prefix? So it would be `mpremote connect /dev/ttyACM0 install foo` and `upip.install("micropython-foo")`?

I would vote for making the distribution name (eg. `micropython-foo`) explicit and compulsory in the manifest file (eg. via `metadata` command) and having `mpremote install` also require this name.

> But the more interesting scenario is for example a manifest in a standalone repo (i.e. a board definition) that is either used for freezing or with the `mpremote deploy app_manifest.py` scenario described above. In that case if the depenency is another git repo available locally, then include() is what you want.  For arbitrary dependencies (i.e. in a non-local github repo) this is the purpose of the (not yet implemented) `FILE_TYPE_HTTP` in this PR. i.e. I plan to support `require("github:user/repo")`.

Would this allow requesting a specific version of the dependency? What about GitLab, Bitbucket and others? Why not something more general like `'some-pkg @ git+https://github.com/someorgname/pkg-repo-name@v1.1'` (taken from a setuptools example)?

> Initially I considered running proxy service that essentially MITMed the request to PyPI. But we really don't want to commit to running any infrastructure.

It looks like micropython.org already has some dynamic webpages (eg https://micropython.org/download/?port=stm32). Adding package conversion to the index could be done quite efficiently performance-wise -- the fact that PyPI doesn't permit replacing uploaded files allows for straightforward on-disk caching of the conversion results. Regarding the development effort -- supporting sdists would probably be too hard and not worth it but converting a whl seems to be not too difficult.

> But I guess I come back to...what's the point of PyPi in that scenario?

The first point is that PyPI and wheel is something people already know. If you supported this approach, the person already familiar with Python packaging could create their module.py and setup.py (or more recent alternative), upload the wheel to PyPI and be done with it -- no need to learn the MicroPython way.

Another point is the namespace. It wouldn't matter so much if each package was independent, but there has to be a clear way of specifying dependencies. An URL (or github project ID) instead of a registered name is a good option but I understand that you intend to support name-based identification as well. In this case I don't see any clearer option than relying on PyPI (unless you want to build your own registry which allows 3rd party contributions). Given the issues with some PyPI `micropython-` names, having the tools first consult micropython.org/pi and fall back to PyPI would not complicate the picture too much IMO.

I understand how mainfest.py can simplify things for micropython-lib but not how would it benefit the 3rd party library creator. If the constrained environment of upip was being taken care of by the conversion proxy -- what would be the remaining issues with wheels-at-PyPI approach?

andrewleech commented:
> I understand how mainfest.py can simplify things for micropython-lib but not how would it benefit the 3rd party library creator. If the constrained environment of upip was being taken care of by the conversion proxy -- what would be the remaining issues with wheels-at-PyPI approach?

99% of my use cases as a professional developer using micropython every day relies on manifests. I don't use upip or the indexes it refers to. So for me, the manifest system working to directly specify and link together dependencies in a project with all its libraries in submodules, producing a binary with everything frozen in it, is the most important system to work reliably and fast.

I want a manifest file in every library repo I create, such that every application I create can have a manifest file that points to all the library ones and the same single simple file format can be used everywhere to weave large applications together. The library repo manifest files will work equally well directly referenced in a submodule or as the target url in a upip install.

> The first point is that PyPI and wheel is something people already know.

I would actually argue against this. I'm my experience very very few python users are more than peripherally aware of this. 

Users know `pip` or one it's friends `pipenv`, `poetry` etc. and how to use it to get a package into their python installation. They don't necessarily know or care that it comes from pypi, and they definitely don't know there's a wheel file it's downloading as an intermediary format unless the installation fails because the wheel package isn't installed so it's fallen back to sdist and then fails to compile and produces a wall of error text.

Even package developers, most of the time they've just copied some other projects setup.py and tweaked it then used one of the myriad of tools like `twine`, `poetry` or `flit` to upload it. Again, the fact it's built a couple of different package formats and uploaded them to pypi is a technicality that just allows them to put `pip install foo` in their readme as install instructions.

Then you get to the advanced package developers who are comfortable writing setuptools hooks for their multi-platform binary bundled modules. Sure, we understand it inside and out, know all about wheels, eggs and sdist etc. The other thing we know about is rolling with changes and using new tools because the latest and greatest recommendation out of pypa changes every year or so and we try to keep up.

I'd say this fragmentation and proliferation of tools is direct proof that users are not familiar with pypi and wheels etc - because it all changes so much there's rarely any one thing to get familiar with. 
While this in itself is not a reason to create further fragmentation, it is an argument to not bend over backwards and burn (non existent) developer spare time to try to hang off a system that's not built for us and is just as likely to change and break compatibility the second we release.

That being said I'm not opposed to wrapper tools being built to generate a setup.py or pyproject.toml automatically from the manifest, then build and upload, then have the proxy to upip install from pypi. But having this system officially supported would really mean micropython _needs_ someone on staff who can maintain and fix it with immediate priority when pypi changes something that breaks the system.

It's just a very heavyweight / with lots of moving parts alternative to a statically generated index that upip can talk directly to. If upip (on desktop at minimum) can also install directly from arbitrary URLs then that really should cover all the same use cases... other than maybe direct install from arbitrary URLs on device... that might be asking too much perhaps.

jimmo commented:
I've updated this PR with the version handling described above. This matches the usage already in https://github.com/micropython/micropython-lib/pull/506

 - manifest.py can now call `metadata(version=, description=, license=).
 - Freezing now automatically injects `__version__` into the module during compilation.
 - micropython-lib compilation will do the same, and also allow fetching specific versions of packages.
 - micropython-lib's index json now includes the metadata.

The script to publish the micropython-lib static content and the updated upip.py are also mostly implemented but I want to focus first on just getting the manifest.py updates done.

> Would this allow requesting a specific version of the dependency?

Yes. For micropython-lib packages this is straightforward (e.g. `>>> upip.install('foo', version="1.2.3")`)

For github, you can use the tag/branch/commit in the URL.

> What about GitLab, Bitbucket and others? Why not something more general like 'some-pkg @ git+https://github.com/someorgname/pkg-repo-name@v1.1' (taken from a setuptools example)?

I don't mind extending the format, but as long as whatever the site is can support a http/s url, then it might not be worth the code size. `github:` was just a simple way to support the 99% use case.

> I would vote for making the distribution name (eg. micropython-foo) explicit and compulsory in the manifest file (eg. via metadata command) and having mpremote install also require this name.

I'm really not in favour of this. Perhaps a future script that publishes to pypi could prepend `micropython-`, but I this just seems unnecessary to make the user write `upip.install("micropython-foo")` when literally every package is called "micropython-xyz". As a user, I just want to install, for example, "hmac".

In the spirit of https://github.com/micropython/micropython/issues/9018, if "hmac" is the package you install in CPython, then it should be "hmac" on MicroPython too.

> I understand how mainfest.py can simplify things for micropython-lib but not how would it benefit the 3rd party library creator. If the constrained environment of upip was being taken care of by the conversion proxy -- what would be the remaining issues with wheels-at-PyPI approach?

A significant number of third-party library authors are writing single-file packages with no dependencies (i.e. a driver). They publish a README.md and mylibrary.py to a github repo. The README.md contains:

```
Run `>>> upip.install('github:user/repo/mylibrary.py')` or `$ mpremote install github:user/repo/mylibrary.py`
```

In the case of a multi-file package they will also need to write a json file with a set of file paths and external dependencies in which case their instructions would be:

```
Run `>>> upip.install('github:user/repo/mylibrary.json')` or `$ mpremote install github:user/repo/mylibrary.json`
```

If they really want wide reach (and .mpy compilation) they can publish to micropython-lib. I have also made `upip.install()` optionally take a base URL, so there's also nothing stopping someone running their own micropython-lib style deployment.

stinos commented:
Currently we do packaging of windows port + all files it needs (some micropython-lib modules, custom modules, dlls, arbitrary support files) using custom code, which basically has some sort of 'env' file as input which lists all components which go into a specific build and then depending on the type of component certain files are copied. To illustrate input is something like

```
bin: [/path/to/fileA.dll, /path/to/fileB.dll]
py: [/path/to/fileA.py]
mod: [/path/to/micropython-lib/argparse, /path/to/customlib/logging]
```

and this results in an output diirectory tree like

```
out/
|--bin
|   |--fileA.dll
|   |--fileB.dll
|--lib
|   |--fileA.py
|   |--argparse.py
|   |--logging
|       |--__init__.py
|       |--handlers.py
```

Suppose upip has support for local files (say `upip.install('/path/to/mylib.json'), destination='/dest'`) would it be possible to replace all or at least some parts of our custom copying with manifests/upip ?

jimmo commented:
> Suppose upip has support for local files (say upip.install('/path/to/mylib.json', destination='/dest') would it be possible to replace all or at least some parts of our custom copying with manifests/upip ?

@stinos Yes this is one of the use case I had imagined based on the original idea from https://github.com/micropython/micropython/pull/8231

So I think in this case the tool would be mpremote, something along the lines of the `deploy app_manifest.py` example I gave above. In reality the `deploy` command does two things: assemble a local directory, then sync that directory to the device. You just want the first part (and this is exactly what the `manifest` command was in #8231).

The missing part for your specific use case is the ability to reference a non-Python file (e.g. the .dll files) in the manifest. This is a feature I want anyway.

Another use case for referencing files is that you want access to the file content from Python, but keeping it in ROM. I can't remember where this conversation was (possibly Slack) but there was a discussion around integrating the functionalty of @peterhinch's `micropython_data_to_py` tool (see https://github.com/micropython/micropython/issues/5394#issuecomment-562918657) into the manifest system.

Unrelated to your use case, but related to being able to reference arbritrary files, the other feature I'm keen to add to manifests is the ability to build a filesystem image which could be merged with the regular firmware to create a default filesystem with the specified files.

dpgeorge commented:
This is a good improvement of the existing manifest system, regardless of how packages are provided online (eg via PyPI or not).

Merged.
